pkg/daemon/ceph/client/mirror_health.go

```diff
@@ -0,0 +1,217 @@
+	c := &mirrorChecker{
+		context:        context,
+		interval:       &defaultHealthCheckInterval,
+		clusterInfo:    clusterInfo,
+		namespacedName: namespacedName,
+		client:         client,
+		monitoringSpec: monitoringSpec,
+		objectType:     object,
+	}
+
+	// allow overriding the check interval
+	checkInterval := monitoringSpec.StatusCheck.Mirror.Interval
+	if checkInterval != nil {
+		logger.Infof("mirroring status check interval for pool/radosnamespace %q is %q", namespacedName.Name, checkInterval.Duration.String())
+		c.interval = &checkInterval.Duration
+	}
+
+	return c
+}
+
+// checkMirroring periodically checks the health of the cluster
+func (c *mirrorChecker) CheckMirroring(context context.Context) {
+	// check the mirroring health immediately before starting the loop
+	err := c.CheckMirroringHealth()
+	if err != nil {
+		c.UpdateStatusMirroring(nil, nil, nil, err.Error())
+		logger.Debugf("failed to check pool/radosnamespace mirroring status for pool/radosnamespace %q. %v", c.namespacedName.Name, err)
+	}
+
+	for {
+		select {
+		case <-context.Done():
+			logger.Infof("stopping monitoring pool/radosnamespace mirroring status %q", c.namespacedName.Name)
+			return
+
+		case <-time.After(*c.interval):
+			logger.Debugf("checking pool/radosnamespace mirroring status %q", c.namespacedName.Name)
+			err := c.CheckMirroringHealth()
+			if err != nil {
+				c.UpdateStatusMirroring(nil, nil, nil, err.Error())
+				logger.Debugf("failed to check pool/radosnamespace mirroring status for ceph block pool/radosnamespace %q. %v", c.namespacedName.Name, err)
+			}
+		}
+	}
+}
+
+func (c *mirrorChecker) CheckMirroringHealth() error {
+	// Check mirroring status
+	mirrorStatus, err := GetPoolMirroringStatus(c.context, c.clusterInfo, c.monitoringSpec.Name)
+	if err != nil {
+		c.UpdateStatusMirroring(nil, nil, nil, err.Error())
+	}
+
+	// Check mirroring info
+	mirrorInfo, err := GetPoolRadosNamespaceMirroringInfo(c.context, c.clusterInfo, c.monitoringSpec.Name)
+	if err != nil {
+		c.UpdateStatusMirroring(nil, nil, nil, err.Error())
+	}
+
+	// If snapshot scheduling is enabled let's add it to the status
+	// snapSchedStatus := cephclient.SnapshotScheduleStatus{}
+	snapSchedStatus := []cephv1.SnapshotSchedulesSpec{}
+	if c.monitoringSpec.Mirroring.SnapshotSchedulesEnabled() {
+		snapSchedStatus, err = ListSnapshotSchedulesRecursively(c.context, c.clusterInfo, c.monitoringSpec.Name)
+		if err != nil {
+			c.UpdateStatusMirroring(nil, nil, nil, err.Error())
+		}
+	}
+
+	// On success
+	c.UpdateStatusMirroring(mirrorStatus.Summary, mirrorInfo, snapSchedStatus, "")
+
+	return nil
+}
+
+// updateStatusBucket updates an object with a given status
+func (c *mirrorChecker) UpdateStatusMirroring(mirrorStatus *cephv1.MirroringStatusSummarySpec, mirrorInfo *cephv1.MirroringInfo, snapSchedStatus []cephv1.SnapshotSchedulesSpec, details string) {
+	switch c.objectType.(type) {
+	case *cephv1.CephBlockPool:
+		blockPool := &cephv1.CephBlockPool{}
```

Comment:
For readability/maintainability, I'd suggest to factor the contents of each case into helper methods such as `UpdatePoolStatusMirroring()` and `UpdateRadosNamespaceStatusMirroring()`.